VirtoCommerce.Platform.Web/Controllers/ExternalLoginController.cs

```diff
@@ -0,0 +1,200 @@
+
+                case SignInStatus.Failure:
+                    await CreateOrUpdatePlatformUser(externalLoginInfo, userName, signInManager);
+                    return Redirect(returnUrl);
+
+                case SignInStatus.LockedOut:
+                case SignInStatus.RequiresVerification:
+                    // TODO: handle user lock-out and two-factor authentication
+                    return Redirect(returnUrl);
+
+                default:
+                    throw new InvalidOperationException($"External login result has the unexpected value: {externalLoginResult}.");
+            }
+        }
+
+        private async Task SignInExternalUser(ExternalLoginInfo externalLoginInfo, string userName, ApplicationSignInManager signInManager)
+        {
+            var userManager = _userManagerFactory();
+
+            var platformUser = await _securityService.FindByNameAsync(userName, UserDetails.Reduced);
+            if (platformUser == null)
+            {
+                throw new InvalidOperationException($"User '{userName}' exists in ASP.NET Identity database, but there is no corresponding VC platform account.");
+            }
+
+            var aspnetUser = await userManager.FindByNameAsync(userName);
+            if (aspnetUser == null)
+            {
+                throw new InvalidOperationException($"User '{userName}' does not have an ASP.NET Identity account.");
+            }
+
+            await signInManager.SignInAsync(aspnetUser, true, true);
+            await _eventPublisher.Publish(new UserLoginEvent(platformUser));
+        }
+
+        private async Task CreateOrUpdatePlatformUser(ExternalLoginInfo externalLoginInfo, string userName,
+            ApplicationSignInManager signInManager)
+        {
+            var platformUser = await _securityService.FindByNameAsync(userName, UserDetails.Reduced);
+            if (platformUser != null)
+            {
+                await AddExternalLoginToExistingUser(platformUser, externalLoginInfo);
+            }
+            else
+            {
+                platformUser = await RegisterExternalUser(userName, externalLoginInfo);
+            }
+
+            var userManager = _userManagerFactory();
+            var aspnetUser = await userManager.FindByNameAsync(userName);
+            if (aspnetUser == null)
+            {
+                throw new InvalidOperationException($"ASP.NET Identity account for user '{userName}' could not be found.");
+            }
+
+            await signInManager.SignInAsync(aspnetUser, true, true);
+            await _eventPublisher.Publish(new UserLoginEvent(platformUser));
+        }
+
+        private async Task AddExternalLoginToExistingUser(ApplicationUserExtended platformUser,
+            ExternalLoginInfo externalLoginInfo)
+        {
+            var externalLogins = platformUser.Logins?.ToList() ?? new List<ApplicationUserLogin>();
+
+            var externalLogin = ConvertExternalLoginInfoToExternalLogin(externalLoginInfo);
```

Comment:
Are you sure what this logic won't create duplicate records after each external login?

Reply:
Well, it shouldn't - if the user already had an external login entry, they will sign in successfully (`externalLoginResult == SignInStatus.Success` on line 83), and this method won't get called. And even if it *will* create duplicate record and attempt to save it, it will lead to duplicate record in `dbo.AspNetUserLogins` table - such record will violate the primary key constraint, and RDBMS will not allow to save it.
![image](https://user-images.githubusercontent.com/1835759/45994632-b6eb6700-c0be-11e8-9e5b-88c21643247b.png)

Reply:
Reworked the code that adds external logins. Now there won't be any external login duplications.